### PR TITLE
Group API requests

### LINKED ISSUE #192: Arisa adds a single changelog entry per action
## The Problem
Arisa adds a single changelog entry for every single action that it takes. That can get quite spammy at times.

This affects every module that does more than one action as far as I can tell.

## Example ticket
https://bugs.mojang.com/browse/MC-159283

## Possible Solution
Group API requests. This is possible because we as moderators can do that via the website. This is probably not easy to implement.

### PR DESCRIPTION
**Pull request by SPGoding**
## Purpose
Fix #192.
## Approach
Added a new cache object named `IssueUpdateContextCache`, which stores a map from issue keys to the necessary information (stored in `IssueUpdateContext`) we need to make an update and a transition. Functions in `infrastructure.jira.Functions` now accept the `IssueUpdateContext` class instead of the `JiraIssue` class.

At the end of each run we will apply the updates and transitions of every `IssueUpdateContext` stored in the cache, and then clear the cache.

Changes that use the same update method and are run in the same run will be grouped after merging this PR:
| Field / Operation | Update method | Note |
| - | - | - |
| CHK | transition |
| Confirmation Status | transition |
| Linked | transition |
| security | transition |
| resolution | transition |
| reopen | transition |
| resolveAs | transition |
| description | update |
| Affected Version/s | update | Technically this can also be done by transition, however the library doesn't provide `fieldAdd` for transitions. |
| createComment | N/A | Technically this can also be done by transition, however the library doesn't provide `fieldAdd` for transitions. |
| addRestrictedComment | N/A | Technically this can also be done by transition, however the library doesn't provide `fieldAdd` for transitions. |
| deleteAttachment | N/A |
| createLink | N/A |
| deleteLink | N/A |
| updateCommentBody | N/A |
| restrictCommentToGroup | N/A |
| getGroups | N/A |
#### Checklist
- [ ] Included tests
- [ ] Tested in MCTEST-xxx

### CODE REVIEW

File: src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContextCache.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package io.github.mojira.arisa.domain
+
+import io.github.mojira.arisa.infrastructure.Cache
+
+object IssueUpdateContextCache : Cache<IssueUpdateContext>()
```
Comment (NeunEinser) [inline src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContextCache.kt:5]: Why is this needed?

File: src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContext.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package io.github.mojira.arisa.domain
+
+import net.rcarz.jiraclient.Issue
+
+data class IssueUpdateContext(
+    val update: Issue.FluentUpdate,
+    val transition: Issue.FluentTransition,
+    var hasUpdates: Boolean = false,
+    var hasTransitions: Boolean = false,
+    var transitionName: String = "Update Issue"
```
Comment (NeunEinser) [inline src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContext.kt:10]: This currently does only support one transition.
It is conceivable that we will need both an update and another transition at some point though.

File: src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt [hunk 3]
```diff
@@ -18,48 +20,93 @@ import java.net.URI
...
+}
+
+fun updateSecurity(context: IssueUpdateContext, levelId: String) {
+    context.hasTransitions = true
+    context.transition.field(Field.SECURITY, Field.valueById(levelId))
+}
+
+fun removeAffectedVersion(context: IssueUpdateContext, version: Version) {
+    context.hasUpdates = true
+    context.update.fieldRemove("versions", version)
```
Comment (NeunEinser) [inline src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:81]: To avoid having both an update transition and a regular update, this could be moved to the update transition instead.
This is true for everything in a regular update, except of the description.

If you'd like to avoid the change being denoted as "Opened", "Reopened" or "Resolved" in emails and logs and and avoid having the status field show up in the changelog, you could do a bit more of an advanced approach and have a registry where fields are regisered with booleans whether it's possible to change them in an update and an update transiton, and prefer an update, when everything is possible as a simple update.

File: src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt [hunk 4]
```diff
@@ -55,6 +57,9 @@ class ModuleExecutor(
                     )
                 }
 
+                IssueUpdateContextCache.forEach(::applyUpdatesAndTransitions)
+                IssueUpdateContextCache.clear()
```
Comment (NeunEinser) [inline src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt:61]: This is currently in the executeModules method, which means that in case one function or one module does something wrong, the entire run will fail and last-run will remain at the old time.

This should move into each module's execute function and instead return a failed module response for that module, so that we only save issue keys where modules actually failed and don't have to run through all issues if something bad happened.

File: src/main/kotlin/io/github/mojira/arisa/infrastructure/Cache.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package io.github.mojira.arisa.infrastructure
+
+@Suppress("MemberNameEqualsClassName")
```
[the remark this answers is not in the document]
Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/Cache.kt:3]: Rip my vocabulary.

File: src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContextCache.kt (hunk 1 above)
Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContextCache.kt:5]: Because I don't feel like I want to pass the cache instance to every mapper function, so I just create it as a singleton.

File: src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt (hunk 4 above)
Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt:61]: Then that will not group these API requests if I understand you correctly

File: src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt (hunk 3 above)
Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:81]: :cry: Sadly the Jira library we use doesn't provide a `fieldAdd` and `fieldRemove` method for transition, as I noted in the PR summary

File: src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContextCache.kt (hunk 1 above)
Comment (urielsalis) [reply src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContextCache.kt:5]: I think its better to pass it, that way we can swap implementations in a single place instead of multiple

File: src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt (hunk 4 above)
Comment (NeunEinser) [reply src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt:61]: That's exactly what happens when one module fails.
If you can adapt that code to collect failed tickets here, that would be fine as well.

File: src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt (hunk 3 above)
Comment (NeunEinser) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:81]: This lib also doesn't support deleting attachments, we had to write our own API call for that.
Does the official one not provide the rest client?

Also I think fieldAdd and fieldRemove could still work by setting the field to all previous entries plus (or minus for remove) one entry. Would be kinda annoying to handle for multiple adds/removes, though.

File: src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContext.kt (hunk 2 above)
Comment (NeunEinser) [reply src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContext.kt:10]: Wdym?
The regular updates, yes.
But we have an update transition here. That's a custom transition on our tracker and has nothing to do with the normal update.

The normal update is just a put on the issue, while the update transition is something custom.

Comment (NeunEinser) [reply src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContext.kt:10]: Oh do you mean that you can add fields tro any transition?

You can only add fields to a transition that are available on the screen, otherwise you get a 400 (or something).

For resolve the screen only contains what's visible here:
image

The update transition is the only one that supports these fields.

File: src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt (hunk 3 above)
Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:81]: No, the official one doesn't even expose the rest client (or at least I can't find it anywhere)

File: src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContext.kt (hunk 2 above)
Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/domain/IssueUpdateContext.kt:10]: Good to know!

File: src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt (hunk 3 above)
Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:81]: > If you'd like to avoid the change being denoted as "Opened", "Reopened" or "Resolved" in emails and logs and and avoid having the status field show up in the changelog

iirc updates don't trigger any email, so we don't even to worry about that, right? And I personally don't really care about that status change shows in the changelog since they are anyways in the same item.

Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt]: > Also I think fieldAdd and fieldRemove could still work by setting the field to all previous entries plus (or minus for remove) one entry.

I gave up lol

Comment (NeunEinser) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt]: > 
> 
> > If you'd like to avoid the change being denoted as "Opened", "Reopened" or "Resolved" in emails and logs and and avoid having the status field show up in the changelog
> 
> iirc updates don't trigger any email, so we don't even need to worry about that, right? And I personally don't really care about that status change shows in the changelog since they are anyways in the same item.

Yeah, both methods have advantages and disadvantages, that's why I don't have really have a preference.

Comment (NeunEinser) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt]: Actually, if we want to add versions in update this, it would probably be the easiest to just use our own fork from the (current) lib and implement these things there, instead of constantly working around them

File: src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt [hunk 6]
```diff
@@ -18,142 +23,178 @@ import java.net.URI
...
+    val resolutionJson = JSONObject()
+    resolutionJson["name"] = resolution
 
-fun reopenIssue(issue: Issue) = runBlocking {
-    Either.catch {
-        issue.transition().execute("Reopen Issue")
-    }
+    context.resolve.field(Field.RESOLUTION, resolutionJson)
+    context.transitionName = "Resolve Issue"
 }
```
Comment (NeunEinser) [inline src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:73]: You are executing everything from all modules after a run, right?
What happens if there are two conflicting resolutions (e.g. Duplicate from CrashModule and Invalid from PiracyModule)?

Comment (SPGoding) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:73]: The one that is changed later will take effects, I believe

Comment (NeunEinser) [reply src/main/kotlin/io/github/mojira/arisa/infrastructure/jira/Functions.kt:73]: Looked at the lib code, seems to be a regular map, so the last put should indeed be the one that takes effect.

We might want to prefer Dupe over everything else, as that's usually the most helpful, but since it probably still does the linking and post all the comments, it should really be fine.
We can always change it

**Comment by Marcono1234** [comment]
The build failure after the merge seems to be caused by JitPack: https://github.com/jitpack/jitpack.io/issues/4226
